File: backtracking_rat_maze.cpp
#include <iostream>
#include <vector>

using namespace std;

// Solves 4x4 Maze configuration
#define N 4 

bool isSafe(vector<vector<int>>& maze, int x, int y) {
    if (x >= 0 && x < N && y >= 0 && y < N && maze[x][y] == 1) {
        return true;
    }
    return false;
}

bool solveMazeUtil(vector<vector<int>>& maze, int x, int y, vector<vector<int>>& sol) {
    if (x == N - 1 && y == N - 1 && maze[x][y] == 1) {
        sol[x][y] = 1;
        return true;
    }

    if (isSafe(maze, x, y)) {
        if (sol[x][y] == 1) return false;

        sol[x][y] = 1;

        if (solveMazeUtil(maze, x + 1, y, sol)) return true;
        if (solveMazeUtil(maze, x, y + 1, sol)) return true;

        sol[x][y] = 0; // BACKTRACK
        return false;
    }
    return false;
}

int main() {
    cout << "--- DSA Masterclass: Rat in a Maze (C++) ---" << endl << endl;

    vector<vector<int>> maze = {
        {1, 0, 0, 0},
        {1, 1, 0, 1},
        {0, 1, 0, 0},
        {1, 1, 1, 1}
    };

    vector<vector<int>> sol(N, vector<int>(N, 0));

    if (!solveMazeUtil(maze, 0, 0, sol)) {
        cout << "Solution does not exist" << endl;
    } else {
        cout << "Final Solution Path Matrix (1 = Rat's Track):" << endl;
        cout << "---------------------------------------------" << endl;
        for (int i = 0; i < N; i++) {
            cout << "  ";
            for (int j = 0; j < N; j++) {
                cout << sol[i][j] << " ";
            }
            cout << endl;
        }
        cout << "---------------------------------------------" << endl;
    }
    return 0;
}

File: backtracking_rat_maze.py
def is_safe(maze, x, y, n):
    if x >= 0 and x < n and y >= 0 and y < n and maze[x][y] == 1:
        return True
    return False

def solve_maze_util(maze, x, y, sol, n):
    # Base Case
    if x == n - 1 and y == n - 1 and maze[x][y] == 1:
        sol[x][y] = 1
        return True

    if is_safe(maze, x, y, n):
        if sol[x][y] == 1:
            return False

        # Mark tracking point
        sol[x][y] = 1

        # Move Right
        if solve_maze_util(maze, x + 1, y, sol, n):
            return True

        # Move Down
        if solve_maze_util(maze, x, y + 1, sol, n):
            return True

        # BACKTRACK
        sol[x][y] = 0
        return False

    return False

def main():
    print("--- DSA Masterclass: Rat in a Maze (Python) ---\n")

    maze = [
        [1, 0, 0, 0],
        [1, 1, 0, 1],
        [0, 1, 0, 0],
        [1, 1, 1, 1]
    ]
    
    n = len(maze)
    sol = [[0] * n for _ in range(n)]

    if not solve_maze_util(maze, 0, 0, sol, n):
        print("Solution does not exist")
    else:
        print("Final Solution Path Matrix (Python):")
        for row in sol:
            print("  " + " ".join(map(str, row)))

if __name__ == "__main__":
    main()